funsor/tensor.py

```diff
@@ -405,18 +405,40 @@ def backend(self):
             return "numpy"
 
 
-@dispatch(numeric_array)
-def to_funsor(x):
-    return Tensor(x)
-
-
-@dispatch(numeric_array, Domain)
-def to_funsor(x, output):
-    result = Tensor(x, dtype=output.dtype)
-    if result.output != output:
-        raise ValueError("Invalid shape: expected {}, actual {}"
-                         .format(output.shape, result.output.shape))
-    return result
+# TODO move these registrations to backend-specific files
+@to_funsor.register(torch.Tensor)
+@to_funsor.register(np.ndarray)
+@to_funsor.register(np.generic)
+def tensor_to_funsor(x, output=None, dim_to_name=None):
+    if not dim_to_name:
+        output = output if output is not None else reals(*x.shape)
+        result = Tensor(x, dtype=output.dtype)
+        if result.output != output:
+            raise ValueError("Invalid shape: expected {}, actual {}"
+                             .format(output.shape, result.output.shape))
+        return result
+    else:
+        assert output is not None  # TODO attempt to infer output
+        assert all(isinstance(k, int) and isinstance(v[0], str) and isinstance(v[1], Domain)
+                   for k, v in dim_to_name.items())
+        # logic very similar to pyro.ops.packed.pack
+        # this should not touch memory, only reshape
+        # pack the tensor according to the dim => (name, domain) mapping in inputs
+        packed_inputs = OrderedDict()
+        for dim, size in zip(range(len(x.shape) - len(output.shape)), x.shape):
+            if size == 1:
+                continue  # TODO broadcast domain and shape here
+            name, domain = dim_to_name[dim + len(output.shape) - len(x.shape)]
+            packed_inputs[name] = domain if domain.dtype > 1 else bint(size)
```

Comment:
I would expect this to infer size from the underlying array:
```py
if size == 1:
    continue
name = dim_to_name[dim + len(output.shape) - len(x.shape)]
packed_inputs[name] = bint(size)
```
Can you explain the TODO and why `dim_to_name` appears to have not only a name but also a domain?

Reply:
I think you're right that this can be relaxed to just a name with a domain inferred from `size`.  If we decide to support higher-rank `bint` domains in Funsor we'd need to provide a domain explicitly.